Adding a Gradient Background to HorizonPlots

Niphredil123 opened the pull request:
Inspired by [this](https://starplot.notion.site/aaa0dd71c17943f89850c9a8c43ade50?v=67a88dfb649344b5b059bf337545cc00&p=53107f542545412ea0de9e82c0eb72a4&pm=s) Horizon Plot task in the Notion roadmap, I have worked to add the ability to add a gradient background to horizon plots.

## Key Actions
- Added the function `_apply_gradient_background()` to the HorizonPlot class in `horizon.py`.
  -  Takes the bounds of the exisiting GeoAxes.
  - Creates a set of background axes.
  - Creates a colourmap gradient based on the provided list.
  - Adds the gradient to the background axes using `imshow()`.
  - Ensures the zorder of the starplot figure's GeoAxes is above the background axes.
- Added an if loop to `_init_plot()` in the HorizonPlot class such that `_apply_gradient_background()` is applied if a `gradient_preset` parameter is provided when HorizonPlot is instantiated.
- Added a file, `styles/ext/gradient_presets.yml`, containing six pre-set gradient colour gradient lists.
- Added the extension `GRADIENTS` to `styles/extensions.py` so presets can be set using `extensions.GRADIENTS["<preset name>"]`.

## Additional Actions
- Added a HorizonPlot example file. `examples/horizon_gradient.py` containing an example horizon plot with a gradient.
- Added `docs/examples/horizon-gradient.md` to list the example horizon gradient plot in `horizon_gradient.py`.
- Added an example card of a horizon gradient plot to `docs/examples.md`.

## Adding a Gradient to a HorizonPlot
To create a HorizonPlot with a gradient, one includes the parameter `gradient_preset` when instansiating `HorizonPlot`. This can be done in two ways:

```
p = HorizonPlot(
    altitude=(0, 60),
    azimuth=(135, 225),
    lat=55.079112,  # Stonehaugh, England
    lon=-2.327469,
    dt=dt,
    style=style,
    resolution=4000,
    scale=0.9,
    gradient_preset=extensions.GRADIENTS["<preset name>"]    # presets = bold_sunset, civil_twilight, nautical_twilight, astronomical_twilight, true_night, pre_dawn
)
```
OR
```
p = HorizonPlot(
    altitude=(0, 60),
    azimuth=(135, 225),
    lat=55.079112,  # Stonehaugh, England
    lon=-2.327469,
    dt=dt,
    style=style,
    resolution=4000,
    scale=0.9,
    gradient_preset=[[0.0, '#0B0C40'], [0.3, '#191970'], [0.6, '#4169E1'], [0.8, '#87CEEB'], [1.0, '#F0F8FF']]
)
```
When adding the parameter as a list, the position values must range from 0-1.  
  
Note that `backgound_color` must be set as transparent using a hex with alpha string (e.g. `#ffffff00`) when extending the `PlotStyle()`. This is because the gradient is plotted under the starplot plot, so will not be visible if the background is not transparent.   

## Images
Below are the images I make reference to in `docs/examples/horizon-gradient.md` (first two) and `docs/example.md` (last one).  

<img width="2240" height="1690" alt="horizon_gradient" src="https://github.com/user-attachments/assets/e0c10cfb-de46-462c-9522-50e0efb4bcc8" />  
  
<img width="1323" height="745" alt="gradient_preset_examples" src="https://github.com/user-attachments/assets/3850ec24-0196-4e36-8272-69e0d8dce70c" />  
  
<img width="280" height="211" alt="horizon_gradient-sm" src="https://github.com/user-attachments/assets/95a3cc61-f42f-4869-a55b-e77462d6c048" />

steveberardi commented:
@Niphredil123 Amazing, thanks so much for this contribution!! I've been wanting to add this to Starplot for awhile now, and your example image looks great! And I love that you added those six different gradient definitions.

Give me a few days to review this a little (I'm out of town right now), but one thing I'm curious about is if you got those gradients from some known source? Just wanna make sure we can include them as part of the starplot code and if they're from some open source project I'd like to add some reference to it.

Thanks again! :)

Steve

Niphredil123 commented:
@steveberardi I'm glad, you like it! It was a fun challenge to figure out. No rush with the review :)

For the colours, I looked at night sky colour palettes on Google images for inspiration and then used [MakeGradients.app](https://makegradients.app/) to design my own. That is also what I used to generated the little example squares. From their FAQs: "All gradients you create can be freely used in personal or commercial projects with no restrictions or attributions needed."

The only palette I actually used any hex codes from was number 19 from [this](https://creativebooster.net/blogs/colors/sunset-color-palettes) Creative Booster inspiration post. I used their hex codes for magenta (#932885) and orange (#FFAF36) in `bold_sunset` and orange in `nautical_twilight`. The post is written with the intent to provide inspiration for sunset colours and I didn't use a full palette, but I am happy to change the hex codes if you would prefer. Anything else is a coincidence. 

I found the Notion board in the [contributing](https://github.com/steveberardi/starplot/blob/main/CONTRIBUTING.md) doc. I did find the Trello board later, but referenced the Notion board becuase that was where the gradients were specifically mentioned. 

Hope it all runs smoothly when you check it :)

steveberardi commented:
I finally had a chance to start looking at the code here, and try running it, and noticed a few things so far:

- The gradient background adds a significant amount of time to creating the plot (2+ seconds on my machine). I think the biggest contributing factor to this is the [resolution specified here](https://github.com/steveberardi/starplot/pull/155/files#diff-02390f0417858c110550d6d28f024e0c6ba064d6b2970e70384dfd51d5ab1e14R444). If I divide that number by 2, it runs about 2x faster and the result looks almost just as good 😀 I’m also wondering if we’d get some speedup by rendering the gradient in Pillow vs Matplotlib, or if there's some other way to speed it up
- Ideally, I think it’d be nice to make the gradient definitions part of the existing `background_color` style property. In other words, make that property accept either a color string OR a list of tuples for the gradient. This would be similar to the `line_style` property ([see code here](https://github.com/steveberardi/starplot/blob/main/src/starplot/styles/base.py#L364)). Then, in the `_init_plot` function if the background color is a list of tuples it would draw the gradient
- It would also be awesome to make zenith and optic plots support these gradients, but instead of a vertical gradient for those make it radial from the center

Would you be interested in helping explore solutions to any of the above? If not, no worries, I can also merge this PR as-is and work on them.

Also, thanks so much for taking the time to read the Starplot code and following so many of the current patterns in the code! 😃

Niphredil123 commented:
I worked out how to plot radial gradients as discussed, but when it came to adding it to MapPlot and OpticPlot, I realised that there would be a lot of repeated code. Thus, I decided to move from individual gradient methods in each plot class, to a separate mixin class, called `GradientBackgroundMixin`.

## GradientBackgroundMixin

- The class is located in `/plotters/gradients.py` and has been add added to the directory's `__init__.py` file.
- It can be inherited by `MapPlot`, `HorizonPlot`, and `OpticPlot` in the same manner as the other `/plotters` mixins.
- It has one public method, `apply_gradient_background`, and several supporting protected methods that work together to add a radial, vertical, or mollweide gradient to plots.
- `apply_gradient_background` is called in the `_init_plot(self)` method of each plot class.
- Most of the protected methods are used with the class but `_find_gradient_shape` is overridden by the child Plot class to determine what gradient shape is appropriate.
- The function for creating a mollweide gradient, `_create_mollweide_gradient(self)`, is a little maths heavy. It could probably be done more elegantly with `astropy` but I didn't want to add a new library.

## Changes to Plot Classes

- Each plot class inherits the mixin and declares a variable `self.gradient_preset` when instantiated.
- The inherited method `_find_gradient_shape` is overridden by each Plot class.
    - HorizonPlot returns a gradient shape of `vertical`.
    - MapPlot returns a gradient shape based on the plot projection: `radial` for orthographic, stereographic, and zenith; `mollweide` for mollweide, and `vertical` for anything else.
    - OpticPlot returns `radial` for all but the camera `self.optic.label`, for which it returns a `vertical` gradient shape.
- An `if` statement calling `apply_gradient_background` if there is a `gradient_preset` is added to each plot class's `_init_plot(self)` method.

## Pain Points

- The radial gradient plots use a magic number (`1.11`) to increase the padding to match the amount of empty space around the zenith and circular optic plots.
- While I was able to get most of the plots to work, I haven't been able to get the camera optic gradients to be clipped in a satisfactory manner. The non-rotated camera optic scrapes by, but relies heavily on magic numbers to resize the background axes (where the gradient is) to match the border of the optic.
- This is still working with a separate gradient preset instead of a `background_color` tuple like `line_style`. Please let me know if you'd like me to work towards making that shift, though it make take a while as I have a few big commitments at the moment.

## Additional Actions

- I added some radial gradient presets to the `gradient_preset` file because I found the colour positions of the vertical gradients looked bad when used radially. The colours used are very similar; however, I did have to simplify a couple of them for the best look.
- I also added a couple of presets for the mollweide gradient. I'll confess to not being overly fond of it in general. My original hope was to add some more dynamism to the milky way plots, but I worry this falls rather short of that goal.
- I've added hash checks for:
    - A zenith plot to `zenith_checks.py`; and
    - A mollweide plot and a miller projection plot to `map_checks.py`.
- I added a line to `docs/examples/horizon-gradient.md` noting that gradients can be added to `MapPlot` and `OpticPlot` plots in the same manner, and detailed how to add a custom gradient.
- Time-wise, I found that plotting vertical gradients adds little to no time, while radial gradients add 1-2 seconds on my machine.

## Images

Below are the new radial and mollweide gradients:  

<img width="1326" height="744" alt="gradient_preset_examples_radial" src="https://github.com/user-attachments/assets/6f881f07-ebb6-4b30-8cc0-8bf36b1e906b" />  

<img width="1323" height="536" alt="gradient_preset_examples_mollweide" src="https://github.com/user-attachments/assets/ae03f72f-71ce-4815-8937-664bf504a833" />  
  
And examples of a few of the new plots:  

<img width="1000" height="513" alt="milky_way_mollweide_stars_sm" src="https://github.com/user-attachments/assets/7b7d13ac-cb0e-4262-97ea-9c486e84320b" />  
<img width="1154" height="1106" alt="optic_gradient_example_sm" src="https://github.com/user-attachments/assets/c7cfea88-d49e-47e5-98c3-9651e2037597" />  
<img width="996" height="995" alt="zenith_gradient_example_sm" src="https://github.com/user-attachments/assets/b0bec041-82ae-4616-9cdc-cc7864f46c0f" />  




Apologies for the essay. I hope it all makes sense.
